network/p2p/scoring/subscription_validator.go

```diff
@@ -1,47 +1,52 @@
 package scoring
 
 import (
-	"fmt"
-
 	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/rs/zerolog"
 
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
 	"github.com/onflow/flow-go/network/p2p"
+	"github.com/onflow/flow-go/network/p2p/p2plogging"
 	p2putils "github.com/onflow/flow-go/network/p2p/utils"
 )
 
 // SubscriptionValidator validates that a peer is subscribed to topics that it is allowed to subscribe to.
 // It is used to penalize peers that subscribe to topics that they are not allowed to subscribe to in GossipSub.
 type SubscriptionValidator struct {
+	component.Component
+	logger               zerolog.Logger
 	subscriptionProvider p2p.SubscriptionProvider
 }
 
-func NewSubscriptionValidator() *SubscriptionValidator {
-	return &SubscriptionValidator{}
-}
+func NewSubscriptionValidator(logger zerolog.Logger, provider p2p.SubscriptionProvider) *SubscriptionValidator {
+	v := &SubscriptionValidator{
+		logger:               logger.With().Str("component", "subscription_validator").Logger(),
+		subscriptionProvider: provider,
+	}
 
-var _ p2p.SubscriptionValidator = (*SubscriptionValidator)(nil)
+	v.Component = component.NewComponentManagerBuilder().
+		AddWorker(func(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+			logger.Debug().Msg("starting subscription validator")
+			v.subscriptionProvider.Start(ctx)
+			<-v.subscriptionProvider.Ready()
```

Comment:
does this component become ready quickly? if not, it's probably worth using a select with the context to allow graceful aborts

Reply:
https://github.com/onflow/flow-go/pull/4988/commits/d83fab5faecfac387c482ca02675be44ddce3fd9